lib/rspec/rails/matchers/have_reported_error.rb

```diff
@@ -0,0 +1,272 @@
+            elsif @expected_error
+              return "Expected error to be an instance of #{@expected_error}, but got: #{reported_errors}"
+            elsif @expected_message.is_a?(Regexp)
+              return "Expected error message to match #{@expected_message}, but got: #{reported_errors}"
+            elsif @expected_message.is_a?(String)
+              return "Expected error message to be '#{@expected_message}', but got: #{reported_errors}"
+            end
+          else
+            if @expected_error && !actual_error.is_a?(@expected_error)
+              return "Expected error to be an instance of #{@expected_error}, but got #{actual_error.class} with message: '#{actual_error.message}'"
+            elsif @expected_message
+              return "Expected error message to #{@expected_message.is_a?(Regexp) ? "match" : "be" } #{@expected_message}, but got: '#{actual_error.message}'"
+            else
+              return "Expected specific error, but got #{actual_error.class} with message: '#{actual_error.message}'"
+            end
+          end
+        end
+
+        def failure_message_when_negated
+          warn_about_negated_with_qualifiers! if has_qualifiers?
+
+          error_count = @reports.count
+          error_word = 'error'.pluralize(error_count)
+          verb = error_count == 1 ? 'has' : 'have'
+
+          "Expected the block not to report any errors, but #{error_count} #{error_word} #{verb} been reported."
+        end
+
+        private
+
+        def error_matches_expectation?
+          return true if @expected_error.nil? && @expected_message.nil? && @reports.count.positive?
+
+          @reports.any? do |report|
+            error_class_matches?(report.error) && error_message_matches?(report.error)
+          end
+        end
+
+        def error_class_matches?(error)
+          @expected_error.nil? || error.is_a?(@expected_error)
+        end
+
+        # Check if the given error message matches the expected message pattern
+        def error_message_matches?(error)
+          return true if @expected_message.nil?
+
+          case @expected_message
+          when Regexp
+            error.message&.match(@expected_message)
+          when String
+            error.message == @expected_message
+          else
+            false
+          end
+        end
+
+        def attributes_match_if_specified?
+          return true if @attributes.empty?
+          return false unless matching_report
+
+          report_context = matching_report.context
+          attributes_match?(report_context)
+        end
+
+        def actual_error
+          @actual_error ||= matching_report&.error
+        end
+
+        def matching_report
+          @matching_report ||= find_matching_report
+        end
+
+        def find_matching_report
+          @reports.find do |report|
+            error_class_matches?(report.error) && error_message_matches?(report.error)
+          end
```

Comment:
What if there were two errors reported, same class, same message, but the attributes (context) only match on the second one, not the first one that `find` finds?

There is some repetition between this method and [those lines](https://github.com/rspec/rspec-rails/pull/2849/files#diff-5095b0a095ba1b0511036361c8930a18ff4367591813576e280b69bb823bf655R141-R143).